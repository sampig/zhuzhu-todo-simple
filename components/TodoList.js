import React from 'react'
import { Alert, Button, FlatList, Platform, StyleSheet, View } from 'react-native'
import i18n from 'util/i18n-utils'
import useTodo from 'hooks/useTodo'
import TodoForm from './TodoForm'
import TodoItem from './TodoItem'

const styles = StyleSheet.create({
  containerButtons: {
    padding: 5,
  },
  containerList: {
    flex: 1,
    display: 'flex',
  },
})

/**
 * Component to display the list of to-dos.
 * @param {object} navigation - the navigation object in route.
 * @returns {node}            - React node
 * @constructor
 */
function TodoList ({ navigation }) {
  const { todos, addTodo, removeTodo, clearTodos } = useTodo()

  /**
   * Delete all to-dos.
   */
  const clearAllTodos = () => {
    if (Platform.OS === 'web') {
      /*global confirm*/
      if (confirm(i18n.t('clearAllText'))) {
        clearTodos()
      }
    } else {
      Alert.alert(
        i18n.t('clearAllTitle'),
        i18n.t('clearAllText'),
        [
          {
            text: i18n.t('cancelButton'),
            style: 'cancel',
          },
          {
            text: i18n.t('clearButton'),
            onPress: () => {
              clearTodos()
            },
          },
        ],
        {
          cancelable: true,
        },
      )
    }
  }

  /**
   * Buttons at the bottom of the page.
   * @returns {node} - React node of buttons
   */
  const buttonsBottom = () => {
    return (
      <View style={styles.containerButtons}>
        <Button
          title={i18n.t('clearAllButton')}
          onPress={clearAllTodos}
        />
      </View>
    )
  }

  /**
   * Return a node of the to-do item.
   * @param {object} item - a to-do object.
   * @returns {node}      - React node
   */
  const renderItem = ({ item }) => (
    <TodoItem todoItem={item} navigation={navigation} removeTodo={removeTodo} />
  )

  return (
    <>
      <TodoForm addNewTodo={addTodo} />
      <FlatList
        style={styles.containerList}
        data={todos}
        renderItem={renderItem}
        keyExtractor={item => item.id}
        initialNumToRender={10} // Render only the first 10 items initially
        windowSize={5} // Adjust the number of items rendered outside the viewport
      />
      {buttonsBottom()}
    </>
  )
}

export default TodoList
