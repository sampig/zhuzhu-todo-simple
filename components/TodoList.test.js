import React from 'react'
import renderer from 'react-test-renderer'
import TodoList from './TodoList'

jest.mock('@expo/vector-icons/MaterialIcons', () => 'Icon')
Date.now = jest.fn(() => 7 * 24 * 60 * 60 * 1000 + 10)
const navigation = { navigate: jest.fn() }
jest.mock('hooks/useTodo', () => ({
  __esModule: true,
  default: () => ({
    todos: [
      { id: 'testId1', title: 'test title 1', createTime: 1 },
      { id: 'testId2', title: 'test title 2', createTime: 100 },
      { id: 'testId3', title: 'test title 3', createTime: 6 * 24 * 60 * 60 * 1000 + 10 },
      { id: 'testId4', title: 'test title 4', createTime: 7 * 24 * 60 * 60 * 1000 },
    ],
    addTodo: jest.fn(),
    removeTodo: jest.fn(),
    clearTodos: jest.fn(),
  }),
}))

describe('The TodoList component', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<TodoList navigation={navigation} />)
    const render = component.toJSON()
    expect(render).toMatchSnapshot()
  })
})
