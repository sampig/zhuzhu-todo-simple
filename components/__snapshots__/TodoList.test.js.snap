// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The TodoList component renders without crashing 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="enter a to-do..."
      rejectResponderTermination={true}
      style={
        Object {
          "borderRadius": 5,
          "borderWidth": 2,
          "flex": 1,
          "fontSize": 20,
          "margin": 5,
          "padding": 5,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
             +   
        </Text>
      </View>
    </View>
  </View>,
  <RCTScrollView
    data={
      Array [
        Object {
          "createTime": 1,
          "id": "testId1",
          "title": "test title 1",
        },
        Object {
          "createTime": 100,
          "id": "testId2",
          "title": "test title 2",
        },
        Object {
          "createTime": 518400010,
          "id": "testId3",
          "title": "test title 3",
        },
        Object {
          "createTime": 604800000,
          "id": "testId4",
          "title": "test title 4",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "display": "flex",
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "rgba(255, 255, 255, 0.8)",
              "borderWidth": 2,
              "flexDirection": "row",
              "margin": 8,
              "padding": 8,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flex": 6,
                  "flexDirection": "row",
                },
                Object {
                  "borderColor": "#ff0000",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "marginLeft": 8,
                  "marginRight": 8,
                  "padding": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontSize": 20,
                    }
                  }
                >
                  test title 1
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Icon
              color="green"
              name="check"
              size={40}
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "rgba(255, 255, 255, 0.8)",
              "borderWidth": 2,
              "flexDirection": "row",
              "margin": 8,
              "padding": 8,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flex": 6,
                  "flexDirection": "row",
                },
                Object {
                  "borderColor": "#cccc00",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "marginLeft": 8,
                  "marginRight": 8,
                  "padding": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontSize": 20,
                    }
                  }
                >
                  test title 2
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Icon
              color="green"
              name="check"
              size={40}
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "rgba(255, 255, 255, 0.8)",
              "borderWidth": 2,
              "flexDirection": "row",
              "margin": 8,
              "padding": 8,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flex": 6,
                  "flexDirection": "row",
                },
                Object {
                  "borderColor": "#0000ff",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "marginLeft": 8,
                  "marginRight": 8,
                  "padding": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontSize": 20,
                    }
                  }
                >
                  test title 3
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Icon
              color="green"
              name="check"
              size={40}
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "rgba(255, 255, 255, 0.8)",
              "borderWidth": 2,
              "flexDirection": "row",
              "margin": 8,
              "padding": 8,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flex": 6,
                  "flexDirection": "row",
                },
                Object {
                  "borderColor": "#007700",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "marginLeft": 8,
                  "marginRight": 8,
                  "padding": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontSize": 20,
                    }
                  }
                >
                  test title 4
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Icon
              color="green"
              name="check"
              size={40}
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    style={
      Object {
        "padding": 5,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Clear all
        </Text>
      </View>
    </View>
  </View>,
]
`;
